### ISSUE TITLE
Can't get it to work with WP 4.9.8

### ISSUE THREAD

**Issue by rmsilver7** [PRDeltoid/Petpoint-Webservices#6]
@PRDeltoid I am building a new WP site for a rescue here in Oregon and I am trying to use your plugin but it is not displaying any data. Could I have something configured incorrectly? I have attached an image of my settings page. I have the div on the appropriate pages and I have refreshed the permalinks, as per the instructions, but still no joy. 

We don't have any "rooms" but I have tried putting in the Location (as pictured) and then the Sublocation and still neither showed anything. 

Any help would be appreciated! Thank you so much for making this available. 

![ppws-config](https://user-images.githubusercontent.com/43621346/46232508-4776b100-c324-11e8-9702-e4ae2812630c.jpg)

**Comment by PRDeltoid** [comment]
Can you try and load this url in your browser: 

> http://ws.petango.com/webservices/wsadoption.asmx/AdoptableSearch?authkey=<pp_auth_key>&speciesID=0&sex=A&ageGroup=ALL&location=Portland%20Fosters&site=Adoptions&onHold=A&orderBy=ID&primaryBreed=All&secondaryBreed=All&specialNeeds=A&noDogs=A&noCats=A&noKids=A&stageid=0

replacing <pp_auth_key> with your Petpoint Authentication Key, and tell me if any results are returned? Everything appears to be hooking and setting up correctly, there are just no results being displayed, so I suspect none are being returned. If you need me to explain this further, let me know.

Looking at the URL, one thing that jumps to mind is I hard-coded the animal's site of "Adoptions" into it. This could possibly be causing your issue, if you do not have a site named "Adoptions" (which, guessing from the fact that you don't have animal rooms, means you mostly shelter animals via fostering them, instead of having a main adoption facility, am I right?)

**Comment by rmsilver7** [comment]
Hi, thanks for getting back to me! You are correct we do not have a sub-location called Adoptions. We are foster based rescue so we have four main locations (Fresno Fosters, Hemet Fosters, PetSmart and Portland Fosters) and then nine sub-locations that we use people's names as. 

Once I change the string to use: &location=Fresno%20Fosters&site=Kathryn%20Coryell to where the one "available" animal is I get the following screen: 
![pp-output](https://user-images.githubusercontent.com/43621346/46236407-73e5f980-c333-11e8-8aa5-fb9430b1ae83.jpg)

**Comment by rmsilver7** [comment]
And just a little more background, we are just getting everything setup in PetPoint and one have one animal in the "available" status. We just got Web Services turned on this morning. Perviously I was getting en error about being unauthorized in the error_log. That is not happening since this morning when they turned something on on their end.

**Comment by rmsilver7** [comment]
So when I change the site string to: &site=0 in the URL that I execute on my browser it returns the result just fine. When I go into the pullanimals.php file and change the &site=Adoptions to &site=0 I still have no joy. I must be missing something? Sorry I am not much of a programmer.

**Comment by rmsilver7** [comment]
Okay I noticed that you are hard coding "http" in the Petango URL, we are using a secure HTTPS site, so I tried changing the URL to have HTTPS in it on pullanimals.php I get the following errors in my error_log

```
[28-Sep-2018 23:29:26 UTC] PHP Warning:  file_get_contents(): SSL operation failed with code 1. OpenSSL Error messages:
error:14090086:SSL routines:ssl3_get_server_certificate:certificate verify failed in /home/[user]/public_html/wp-content/plugins/Petpoint-Webservices-master/pullanimals.php on line 46
[28-Sep-2018 23:29:26 UTC] PHP Warning:  file_get_contents(): Failed to enable crypto in /home/[user]/public_html/wp-content/plugins/Petpoint-Webservices-master/pullanimals.php on line 46
[28-Sep-2018 23:29:26 UTC] PHP Warning:  file_get_contents(https://ws.petango.com/webservices/wsadoption.asmx/AdoptableSearch?authkey=[KEY REMOVED]&amp;speciesID=0&amp;sex=A&amp;ageGroup=ALL&amp;location=Fresno%20Fosters&amp;site=0&amp;onHold=A&amp;orderBy=ID&amp;primaryBreed=All&amp;secondaryBreed=All&amp;specialNeeds=A&amp;noDogs=A&amp;noCats=A&amp;noKids=A&amp;stageid=0): failed to open stream: operation failed in /home/l4drorg/public_html/wp-content/plugins/Petpoint-Webservices-master/pullanimals.php on line 46
[28-Sep-2018 23:29:26 UTC] PHP Warning:  array_merge(): Argument #2 is not an array in /home/[user]/public_html/wp-content/plugins/Petpoint-Webservices-master/pullanimals.php on line 56
[28-Sep-2018 23:29:26 UTC] PHP Warning:  array_filter() expects parameter 1 to be array, null given in /home/[user]/public_html/wp-content/plugins/Petpoint-Webservices-master/pullanimals.php on line 57
```

**Comment by rmsilver7** [comment]
Okay so I worked on this a bit and I got the HTTPS/SSL issues resolved by adding this function to the pullanimals.php file:

```
function file_get_contents_curl( $url ) {

  $ch = curl_init();

  curl_setopt( $ch, CURLOPT_AUTOREFERER, TRUE );
  curl_setopt( $ch, CURLOPT_HEADER, 0 );
  curl_setopt( $ch, CURLOPT_RETURNTRANSFER, 1 );
  curl_setopt( $ch, CURLOPT_URL, $url );
  curl_setopt( $ch, CURLOPT_FOLLOWLOCATION, TRUE );

  $data = curl_exec( $ch );
  curl_close( $ch );

  return $data;

}
```

And then changing line #63 to call the URL with `file_get_contents_curl` 

But I am at a loss of how to fix the array issues, can you provide any help there @PRDeltoid? 

```
[01-Oct-2018 05:51:24 UTC] PHP Warning:  array_merge(): Argument #2 is not an array in /home/[user]/public_html/wp-content/plugins/Petpoint-Webservices-master/pullanimals.php on line 73
[01-Oct-2018 05:51:24 UTC] PHP Warning:  array_filter() expects parameter 1 to be array, null given in /home/[user]/public_html/wp-content/plugins/Petpoint-Webservices-master/pullanimals.php on line 74

```

**Comment by PRDeltoid** [comment]
I think the HTTPS problem is probably a Same-Origin-Policy issue (which I tried to get around with this plugin originally, but I think HTTPS is stricter than HTTP when it comes to pulling data in from other websites). It's awesome that you found a solution, and I'll try to make sense of it to figure out if it's safe to merge into the code (with your permission, of course).

As for the Array_merge/filter issue, I am thinking we are still having issues with the "pull_animal_data" function not receiving any animal data, so the array is "null" (empty) and the functions that want to manipulate that array are throwing an error about it. 

I am in the process of setting up a virtual machine to run a WordPress server so I can actually start testing this issue. Sorry for the delay, I assure you I am still working on this.

**Comment by rmsilver7** [comment]
Awesome thank you for your help, at this point I don’t think there is anything else I can do, but I’m more than happy to test for you (or I can provide you a login to our WP if that will help, let me know). Feel free to use that piece of code, I found it from someone else running into the same issue with a different project on a developers website.

**Comment by PRDeltoid** [comment]
I wasn't able to reproduce an issue with my PetPoint database, so I think the issue lies within that PetPoint URL and the PHP code that we discussed previously. Since I can't troubleshoot the issue on my end, I'll need to try to fix it on your end. If you feel comfortable sharing your WordPress information with me, you can email me at tbritton(AT)sspca.org. This is my work email, so that you can feel more confident in my intentions (more accountability). Giving someone access to your WordPress backend is a big responsibility, and I don't take it lightly. 

There is also the option of sharing your PetPoint API key, which I can use to compose the URL myself and test some different options. After I'm done, you can regenerate the key (you may have to talk to PetPoint to do this) and I will no longer have access. The API Key only allows me to read information out of your PetPoint database, not write/change any information. 

If you're interested in either of these options, please don't respond in this thread with the WP info or API key. Anyone reading this could potentially use it to damage your website or PetPoint Webservices usage (I'm sure you already know this, but I want to be 100% sure).

**Comment by PRDeltoid** [comment]
I've been working on this for a bit and made a bit of progress, but think I've run into an issue with how the site is configured. Could you or the webmaster for your site (if it isn't you) contact me at my email address in the previous message? I would like to discuss the problem and potential solutions.

**Comment by dmkjr** [comment]
@PRDeltoid good evening.  Was following to see if this has been corrected for HTTPS sites.  We are looking to integrate our PetPoint data to our WP site.

**Comment by PRDeltoid** [comment]
I have managed to set up a test site and found that there might have actually been other bugs involved here instead of the `https` portion of that link. I believe the format of the API request to PetPoint did not work for all shelter setups and that one of the values may have been causing your results to come back empty.

This caused another error (the one causing that `array_merge()` error message) because no results were returned due to the above bug. 

I have since fixed both of these bugs. The plugin functions with https websites now and I am able to run a test site with https enabled. That `http` call is done on the backend between the webserver and PetPoint's server. The client never sees or knows about it. This is something I will research fixing using @rmsilver7's fix as a starting point.
